Hands-On Lab: Model Comparison  

● Step 1: Train a decision tree, random forest, SVM, and k-NN on the same train/test split from Day 3.  

● Step 2: Evaluate all four with the same metric (e.g. F1-score) and assemble the results into one comparison 
table.  

● Step 3: Report the random forest's top feature importances and interpret them.    

● Step 4: Identify the best-performing model for this dataset and explain, in Markdown, why it likely won   

In [1]:
from sklearn.datasets import load_breast_cancer # It is a binary classification dataset with Two original classes: malignant and benign

from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

Fetch and prepare the Data:

In [ ]:
Cancer = load_breast_cancer() 

X = Cancer.data


y = (Cancer.target == 0).astype(int) # make it into: malignant = 1, benign = 0

x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# create a pipeline that first scales the features and then fits the logistic regression model (scales the data)
DT = make_pipeline(StandardScaler(), DecisionTreeClassifier())
RF = make_pipeline(StandardScaler(), RandomForestClassifier())
SVM = make_pipeline(StandardScaler(), SVC())
KNN = make_pipeline(StandardScaler(), KNeighborsClassifier())

#DT = DecisionTreeClassifier(random_state=42) (YOU CAN ALSO USE max_depth=3 instead of random_state=42 TO LIMIT THE DEPTH OF THE TREE)
#RF = RandomForestClassifier(n_estimators=100, random_state=42)
#SVM = SVC(kernel='linear', random_state=42)
#KNN = KNeighborsClassifier(n_neighbors=5)


Fit and train the models:

In [14]:
DT.fit(x_train, y_train)
RF.fit(x_train, y_train)
SVM.fit(x_train, y_train)
KNN.fit(x_train, y_train)

DT_pred = DT.predict(x_test)
RF_pred = RF.predict(x_test)
SVM_pred = SVM.predict(x_test)
KNN_pred = KNN.predict(x_test)

Random forest's (RF) top feature importances:

(which features mattered most):

In [15]:

print(RF[-1].feature_importances_) 

'''
For this line: print(RF.feature_importances_)

AttributeError: 'Pipeline' object has no attribute 'feature_importances_'

The error happens because RF is a Pipeline object,
not the RandomForestClassifier itself.
The pipeline doesn't have a feature_importances_ attribute—only the specific Random Forest step inside it does.
To fix this, you need to extract the Random Forest step from the pipeline.
'''

'''
Option 1: Access by Index
    print(RF[-1].feature_importances_)
Option 2: Access by Name
    print(RF.named_steps['randomforestclassifier'].feature_importances_)
'''

[0.04275006 0.01467241 0.04917884 0.05817916 0.00727987 0.00970672
 0.04653323 0.09192505 0.00386663 0.0024761  0.00989185 0.00548507
 0.00634383 0.03168604 0.0036535  0.00253137 0.00523779 0.00389553
 0.00574728 0.00563082 0.08064865 0.02383747 0.10972287 0.11890735
 0.0080922  0.01304338 0.03974817 0.17634802 0.01548098 0.00749979]


"\nOption 1: Access by Index\n    print(RF[-1].feature_importances_)\nOption 2: Access by Name\n    print(RF.named_steps['randomforestclassifier'].feature_importances_)\n"

Feature Importance in Random Forests:   
In Random Forests, feature importance is a measure of how much each feature contributes to the model's predictive power.    

[0.04275006 0.01467241 0.04917884 0.05817916 0.00727987 0.00970672  
 0.04653323 0.09192505 0.00386663 0.0024761  0.00989185 0.00548507  
 0.00634383 0.03168604 0.0036535  0.00253137 0.00523779 0.00389553  
 0.00574728 0.00563082 0.08064865 0.02383747 0.10972287 0.11890735  
 0.0080922  0.01304338 0.03974817 0.17634802 0.01548098 0.00749979] 
 
Interpretation: 
- The output is an array of feature importance scores corresponding to each feature in the dataset.     
- Those with higher values are considered more important for making predictions.    
- Those features are the ones that the Random Forest model relies on more heavily when making decisions.    
in short they are inputs drive predictions most. (The most influencial features in the dataset).


(It is only for tree-based models (SVM, KNN) doesn't have feature importance!)

Classification Reports for the 4 models:

In [21]:

DT_Class_Rep = classification_report(y_test, DT_pred , target_names=['benign', 'malignant'])
RF_Class_Rep = classification_report(y_test, RF_pred , target_names=['benign', 'malignant'])
SVM_Class_Rep = classification_report(y_test, SVM_pred , target_names=['benign', 'malignant'])
KNN_Class_Rep = classification_report(y_test, KNN_pred , target_names=['benign', 'malignant'])


Printing the reports for Decision Tree & Random Forest:

In [22]:

print ("Decision Tree Classifier Report:\n", DT_Class_Rep)
print ("Random Forest Classifier Report:\n", RF_Class_Rep)


Decision Tree Classifier Report:
               precision    recall  f1-score   support

      benign       0.94      0.94      0.94        71
   malignant       0.91      0.91      0.91        43

    accuracy                           0.93       114
   macro avg       0.93      0.93      0.93       114
weighted avg       0.93      0.93      0.93       114

Random Forest Classifier Report:
               precision    recall  f1-score   support

      benign       0.96      0.97      0.97        71
   malignant       0.95      0.93      0.94        43

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



Printing the reports for Support vector machine & k-nearest neighbors:

In [23]:
print ("Support Vector Machine Classifier Report:\n", SVM_Class_Rep)
print ("K-Nearest Neighbors Classifier Report:\n", KNN_Class_Rep)

Support Vector Machine Classifier Report:
               precision    recall  f1-score   support

      benign       0.97      1.00      0.99        71
   malignant       1.00      0.95      0.98        43

    accuracy                           0.98       114
   macro avg       0.99      0.98      0.98       114
weighted avg       0.98      0.98      0.98       114

K-Nearest Neighbors Classifier Report:
               precision    recall  f1-score   support

      benign       0.96      0.96      0.96        71
   malignant       0.93      0.93      0.93        43

    accuracy                           0.95       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.95      0.95      0.95       114




Classification method--maliginant: --f1-score---------recall
     

Decision Tree  ------------------------   0.91 ----------------0.91

Random Forest  ----------------------0.94 ----------------0.93

SVM   --------------------------------- 0.98 ----------------0.95

K-NN --------------------------------  0.93 ----------------0.93

Since this is a medical application, both F1 score and recall are highly important. 

SVM proved to be the best model in both metrics for this case.  

It gave the best results with the highest precision, recall, and f1-score for both classes (benign and malignant).

it is due to its ability to find the optimal hyper_plane that separates Two classes in feature space, even in high-dimensional spaces. 

In non linear cases it can use kernal to transform the data into higher dimensions where it can find a linear separation. 

SVM is less prone to overfitting compared to Decision Trees and KNN.